### PR TITLE
Deprecate lead, trail, lm, lt, lc, valence and coeffs.

### PR DESCRIPTION
**Pull request by wbhart**
I have made PRs against all the other repos.

### CODE REVIEW

**Comment by wbhart** [comment]
Fixes #802 
Fixes #779
Fixes #508

**Comment by wbhart** [comment]
The OscarCI tests are exiting with signal 11 with no stack trace. Any idea what that is about?

**Comment by thofma** [comment]
Very weird, looking at the timestamps does not help either:
```
Wed, 17 Mar 2021 10:19:17 GMT SLPoly        |  414    414
Wed, 17 Mar 2021 10:19:17 GMT┌ Warning: `lead(f)` is deprecated, use `leading_coefficient(f)` instead.
Wed, 17 Mar 2021 10:19:17 GMT│   caller = leading_coefficient(f::fmpq_poly) at Poly.jl:127
Wed, 17 Mar 2021 10:19:17 GMT└ @ Hecke ~/work/AbstractAlgebra.jl/AbstractAlgebra.jl/oscar-dev/Hecke/src/Misc/Poly.jl:127
Wed, 17 Mar 2021 13:13:07 GMT ERROR: Package Oscar errored during testing (received signal: 11)
Wed, 17 Mar 2021 13:13:07 GMT Stacktrace:
```
Let's restart.

**Comment by wbhart** [comment]
There's a lot of "unsatisfiable requirements" (dependencies) as well. I don't know if that caused it somehow.

**Comment by wbhart** [comment]
I have decided to do map_coeffs in a separate series of PR's if we decide to rename that. This is getting too messy otherwise. Thus, I will finish this off as soon as 801 is merged.

**Comment by thofma** [comment]
#801 is merged.

How do we want to move forward? We can merge all the breaking things (+ the to be created PR for map_coeff) today or beginning of next week and immediately tag a new release.

**Comment by fieker** [comment]
On Fri, Mar 19, 2021 at 02:03:00AM -0700, thofma wrote:
> #801 is merged.
> 
> How do we want to move forward? We can merge all the breaking things (+ the to be created PR for map_coeff) today or beginning of next week and immediately tag a new release.
> 
I'd rather do this earlier than later. There will be more of the same I
am sure, but we canot wait and let the PRs get stale.

I can lend a hand if you tell me what to do (only one exam today)

**Comment by wbhart** [comment]
Please do not merge anything else until I have reviewed it. The order of merging here is potentially going to create some massive headaches.

I am on it now and will let you know when everything is ok to go.

**Comment by wbhart** [comment]
I believe this is right to go, independently of everything else. But I have not tested that the deprecations in AA are sufficient for Hecke/Oscar, etc. I'll test Nemo and Hecke now, and if so will call it good. Just give me 30 mins or so to run the tests.

**Comment by wbhart** [comment]
Current failure is just #788

**Comment by wbhart** [comment]
Nemo is fine, lots of yellow warnings, which is what I wanted to see. But tests pass.

Waiting for Hecke tests now.

**Comment by wbhart** [comment]
Hmm, the Hecke tests just seem to have hung without running a single test. 10 mins and counting. Does it do that sometimes?

**Comment by wbhart** [comment]
It seemed to be hung in deprecated.jl

**Comment by thofma** [comment]
Weird. Let me also try it locally.

**Comment by wbhart** [comment]
Looks like the same problem on restart of the tests. It displays:

```
[ Info: long_test: false
[ Info: short_test: false
┌ Warning: `lead(f)` is deprecated, use `leading_coefficient(f)` instead.
│   caller = leading_coefficient(::AbstractAlgebra.Generic.Poly{nf_elem}) at Poly.jl:127
└ @ Hecke ~/.julia/dev/Hecke/src/Misc/Poly.jl:127
```

Then it just hangs. Any idea what is happening here? It seems like the deprecations don't work with Hecke.

**Comment by wbhart** [comment]
Here's the trace when I hit CTRL-C:

```
^CERROR: InterruptException:
Stacktrace:
 [1] poptask(::Number fields: Error During Test at Base.InvasiveLinkedListSynchronized{Task}) at ./task.jl:704
 [2] wait at ./task.jl:712 [inlined]
 [3] wait(::Base.GenericCondition{Base.Threads.SpinLock}) at ./condition.jl:106
 [4] wait(::Base.Process) at ./process.jl:621
 [5] success at ./process.jl:483 [inlined]
 [6] /home/wbhart/.julia/dev/Hecke/test/NumField.jl:1
  Got exception outside of a @test
  LoadError: InterruptException:
  Stacktrace:run
(   [1] backtrace() at ./error.jl:112::
Cmd   [2] macro expansion at ./deprecated.jl:89 [inlined];
   [3] macro expansion at ./logging.jl:321 [inlined]
   [4] depwarn(::String, ::Symbol; force::Bool) at ./deprecated.jl:85
   [5] depwarn(::String, ::Symbol) at ./deprecated.jl:80
   [6] lead(::AbstractAlgebra.Generic.Poly{nf_elem}) at ./deprecated.jl:71
   [7] leading_coefficient(::AbstractAlgebra.Generic.Poly{nf_elem}) at /home/wbhart/.julia/dev/Hecke/src/Misc/Poly.jl:127
   [8] lead(::AbstractAlgebra.Generic.Poly{nf_elem}) at ./deprecated.jl:72
   ... (the last 2 lines are repeated 9389 more times)
   [18787] leading_coefficient(::AbstractAlgebra.Generic.Poly{nf_elem}) at /home/wbhart/.julia/dev/Hecke/src/Misc/Poly.jl:127
   [18788] _preproc_pol(::AbstractAlgebra.Generic.Poly{nf_elem}, ::AbstractAlgebra.Generic.Poly{nf_elem}) at /home/wbhart/.julia/dev/Hecke/src/NumField/NfAbs/Poly.jl:142
   [18789] gcd_modular_kronnecker(::AbstractAlgebra.Generic.Poly{nf_elem}, ::AbstractAlgebra.Generic.Poly{nf_elem}, ::Bool) at /home/wbhart/.julia/dev/Hecke/src/NumField/NfAbs/Poly.jl:179
   [18790] gcd at /home/wbhart/.julia/dev/Hecke/src/NumField/NfAbs/Poly.jl:74 [inlined]
   [18791] issquarefree at /home/wbhart/.julia/dev/Hecke/src/NumField/NfAbs/Poly.jl:432 [inlined]
   [18792] isirreducible_easy(::AbstractAlgebra.Generic.Poly{nf_elem}) at /home/wbhart/.julia/dev/Hecke/src/NumField/NfAbs/Elem.jl:547
   [18793] isirreducible at /home/wbhart/.julia/dev/Hecke/src/NumField/NfAbs/Elem.jl:532 [inlined]
   [18794] #NumberField#1340 at /home/wbhart/.julia/dev/Hecke/src/NumField/NfRel/NfRel.jl:177 [inlined]
   [18795] #NumberField#1341 at /home/wbhart/.julia/dev/Hecke/src/NumField/NfRel/NfRel.jl:183 [inlined]
   [18796] NumberField(::AbstractAlgebra.Generic.Poly{nf_elem}) at /home/wbhart/.julia/dev/Hecke/src/NumField/NfRel/NfRel.jl:183
   [18797] top-level scope at /home/wbhart/.julia/dev/Hecke/test/NumField/Elem.jl:7
   [18798] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [18799] top-level scope at /home/wbhart/.julia/dev/Hecke/test/NumField/Elem.jl:2
   [18800] include(::String) at ./client.jl:457
   [18801] top-level scope at /home/wbhart/.julia/dev/Hecke/test/NumField.jl:2
   [18802] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [18803] top-level scope at /home/wbhart/.julia/dev/Hecke/test/NumField.jl:2
   [18804] include(::String) at ./client.jl:457
   [18805] top-level scope at timing.jl:174
   [18806] top-level scope at /home/wbhart/.julia/dev/Hecke/test/runtests.jl:78
   [18807] include(::String) at ./client.jl:457
   [18808] top-level scope at none:6
   [18809] eval(::Module, ::Any) at ./boot.jl:331
   [18810] exec_options(::Base.JLOptions) at ./client.jl:272
  in expression starting at /home/wbhart/.julia/dev/Hecke/test/NumField/Elem.jl:1
```

**Comment by thofma** [comment]
Yeah, I just saw it locally. There is a recursion:
```
       [6] lead(f::AbstractAlgebra.Generic.Poly{nf_elem})
         @ AbstractAlgebra ./deprecated.jl:71
       [7] leading_coefficient(f::AbstractAlgebra.Generic.Poly{nf_elem})
         @ Hecke ~/.julia/packages/Hecke/qJVuE/src/Misc/Poly.jl:127
       [8] lead(f::AbstractAlgebra.Generic.Poly{nf_elem})
         @ AbstractAlgebra ./deprecated.jl:72
```
I think we can merge as is.

**Comment by wbhart** [comment]
Maybe. I'm trying the Hecke tests with this, the Nemo and Hecke patches all dev'd (so there are no deprecations activated).

Let me check that works first.

**Comment by wbhart** [comment]
The issue is likely that Hecke currently defines leading_coefficient to be lead.

**Comment by thofma** [comment]
I removed the method in Hecke locally and I am now running the tests.

**Comment by wbhart** [comment]
Naturally my Hecke PR already removes this method that is causing the recursion.

**Comment by wbhart** [comment]
Remarkably Hecke passes without any issues with this, the Nemo and Hecke PRs that I made.

I'm surprised about that, but you can't help good luck!

**Comment by thofma** [comment]
Yes, very nice. Feel free to merge #803, #805, #806 in whatever order is the easiest.

**Comment by wbhart** [comment]
Nemo also. We should buy a lottery ticket!

So if you want to go ahead and review and merge this PR, you are welcome to.

Then I will rebase my multivariates on it while you review and merge https://github.com/Nemocas/AbstractAlgebra.jl/pull/806
assuming it doesn't also need a rebase.

File: src/AbstractAlgebra.jl
Comment (thofma) [inline src/AbstractAlgebra.jl:481]:
```suggestion
		   truncate, typed_hcat, typed_hvcat,
```

File: src/AbstractAlgebra.jl
Comment (thofma) [inline src/AbstractAlgebra.jl:534]:
```suggestion
		   left_kernel, leglength, length, lu, lu!,
```

File: src/AbstractAlgebra.jl
Comment (thofma) [inline src/AbstractAlgebra.jl:537]:
```suggestion
		   map_entries!, map_with_preimage_from_func,
```

File: src/AbstractAlgebra.jl
Comment (thofma) [inline src/AbstractAlgebra.jl:569]:
```suggestion
		   truncate, typed_hcat, typed_hvcat,
```

**Review by thofma — APPROVED** [review]

**Comment by wbhart** [comment]
Thanks. I haven't set my VS Code editor up correctly yet. So it still uses tabs equal to 4 spaces and uses actual tabs not spaces. I just didn't find time for that yet. Still so many other problems with my new laptop. :-(

**Comment by wbhart** [comment]
Are all the Oscar CI runs strictly necessary? It looks like there's still about 3 hours of CI before we get an answer on this one!

Can't we cut the Oscar CI runs down to a bare minimum?

**Comment by thofma** [comment]
At the moment the backlog is for the normal "Run tests", which are just the usual AA tests, see https://github.com/Nemocas/AbstractAlgebra.jl/actions. The problem is that it runs the tests on every commit, but does not cancel the old runs. Do you know if these is possible with GitHub Actions @benlorenz?

**Comment by thofma** [comment]
Oh boy, it created a lot of combinations to check. We have to tweak this a bit I think.

**Comment by wbhart** [comment]
Ja

**Comment by thofma** [comment]
I can't tweak it right now. You have tested it locally. Should we just merge?

**Comment by wbhart** [comment]
I'd say so. I've only tested it as far as Nemo and Hecke. I can't test it further in any easy manner right now. In particular I do not guarantee that the patches against Singular and Oscar are correct. They should be carefully reviewed before merging them.

**Comment by wbhart** [comment]
Technically I didn't test against the changes you made that I merged. But if you are confident of no typos...

**Comment by wbhart** [comment]
They were just whitespace anyhow, assuming you didn't delete a comma by accident

**Comment by wbhart** [comment]
I think it is not currently as useful as it could be. For example, it might be useful to know if Singular#deprecate_coeffs, Nemo#deprecate_coeffs, AA#deprecate_coeffs passes, but not with Oscar#master,Hecke#master *if* there are also deprecate_coeffs PRs for those packages.

Somehow I really want to know where in the chain things broke.

**Comment by wbhart** [comment]
Also I guess now that we are maintaining master against master it is not likely that master against releases is going to pass due to package manager things.

**Comment by thofma** [comment]
Yeah, it should probably disable the other tests if it finds a branch in a downstream package. I don't know how easy this is. I will mention it to Benjamin.

**Pull request merged**
